priv/posts/security/20241212210148_epmd-public-exposure.md

```diff
@@ -0,0 +1,114 @@
+- Run a port scan:  
+  ```bash
+  nmap -p 4369 -sV [Your Public IP]
+  ```  
+  If the output shows:  
+  ```
+  PORT     STATE SERVICE
+  4369/tcp open  epmd
+  ```
+  Your EPMD port is exposed.
+
+**For Erlang/Elixir users:**
+- Remove or avoid using `-name` or `-sname` if you don’t need clustering.
+- Bind distribution to a non-public interface with `vm.args`:
+  ```bash
+  -kernel inet_dist_use_interface '{127, 0, 0, 1}'
+  ```
+- Use a firewall or security group rules to restrict external access.
+
+**For RabbitMQ users:**
+- RabbitMQ runs on Erlang and uses EPMD behind the scenes. Even if you never interact directly with Erlang, you may still be affected.
+- Check the RabbitMQ networking documentation to learn how to identify exposed ports and close them:  
+  [https://www.rabbitmq.com/docs/networking#epmd](https://www.rabbitmq.com/docs/networking#epmd)  
+- The Erlang Ecosystem Foundation’s security hardening guide explains what EPMD is, which ports it uses, and how to close them down:  
+  [https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/distribution](https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/distribution)
+
+## What is EPMD and How Does it Work?
+
+Erlang was designed from the ground up for concurrency, distribution, and fault tolerance. To support clustering (distributed Erlang nodes connecting and communicating seamlessly), Erlang relies on EPMD. EPMD runs as a small daemon that keeps track of Erlang nodes and their associated ports. When a node starts with `-name` or `-sname`, it registers with EPMD on the host machine’s port 4369.
+
+In a controlled environment—such as an internal network—this works well. EPMD makes it easy for nodes to discover each other, connect, and form distributed clusters.
+
+**Key points:**
+- EPMD listens on port 4369 by default.
+- It’s assumed that EPMD is only reachable from within a trusted network.
+- Many Erlang/Elixir build tools and release pipelines (like Rebar3, Mix, and Distillery releases) enable EPMD by default.
+
+## Why is Exposing EPMD a Bad Idea?
+
+### Risk of Unauthorized Cluster Access
+
+EPMD itself does not enforce strong authentication. Erlang clustering uses a “cookie” as a shared secret, but this cookie is not intended as a robust security mechanism—it’s more of a sanity check to prevent accidental cross-node connections. If an attacker can guess or brute-force the cookie, they could join your cluster.
+
+Once inside the cluster, the attacker could potentially run arbitrary Erlang Remote Procedure Calls (RPCs), giving them full control of the application and underlying system. While no widespread internet-based attacks on EPMD have been documented publicly, the theoretical risk is real. A known brute-force approach is described here:  
+[https://insinuator.net/2017/10/erlang-distribution-rce-and-a-cookie-bruteforcer/](https://insinuator.net/2017/10/erlang-distribution-rce-and-a-cookie-bruteforcer/)
```

Comment:
Should we point to here as well? https://github.com/gteissier/erl-matter?tab=readme-ov-file

Reply:
Yes I think that is a good idea.